File: scripts/digestFasta.pl
#!/usr/bin/env perl 

use warnings;
use strict;
use Data::Dumper;
use Getopt::Long;
use File::Basename qw/basename/;
use Digest::MD5 qw/md5_base64/;
use Digest::SHA qw/sha1_base64 sha256_base64/;

use version 0.77;
our $VERSION="0.3";

# Make a hashing function that is global. It will reference
# a hashing algorithm.
my $hash_function;

local $0 = basename $0;
sub logmsg{local $0=basename $0; print STDERR "$0: @_\n";}
exit(main());

sub main{
  my $settings={};
  GetOptions($settings,qw(help hash=s force out=s)) or die $!;
  usage() if(!@ARGV || $$settings{help});
  $$settings{out} ||= die "ERROR: need an output directory with --out";
  $$settings{hash}||= "md5";

  # Decide on a hashing algorithm
  if(lc($$settings{hash}) eq "md5"){
    $hash_function = \&md5_base64;
  }
  elsif(lc($$settings{hash}) eq "sha1"){
    $hash_function = \&sha1_base64;
  }
  elsif(lc($$settings{hash}) eq "sha256"){
    $hash_function = \&sha256_base64;
  }
  elsif(lc($$settings{hash}) eq "plaintext"){
    $hash_function = \&uc;
  }
  else{
    die "ERROR: I do not understand --hash $$settings{hash}";
  }

  if(-e $$settings{out} && !$$settings{force}){
    die "ERROR: output folder already exists at $$settings{out}. --force to overwrite.";
  }

  # Make the output directory
  mkdir $$settings{out};
  open(my $refFh,     ">", "$$settings{out}/ref.fasta") or die "ERROR: could not write to $$settings{out}/ref.fasta";
  open(my $allelesFh, ">", "$$settings{out}/alleles.tsv") or die "ERROR: could not write to $$settings{out}/alleles.tsv";

  # allele calling, sequencing tech, assembler
  print $allelesFh join("\t", "# locus","allele","hash-type") ."\n";
  for my $f(@ARGV){
    logmsg "Processing $f";
    digestFasta($f, $refFh, $allelesFh, $settings);
  }

  close $allelesFh;
  close $refFh;

  return 0;
}

sub digestFasta{
  my($f, $refFh, $allelesFh, $settings) = @_;
  

  my $numSeqs = 0;
  open(my $seqFh, "<", "$f") or die "ERROR: could not open $f for reading";
  my @aux = undef;
  my ($id, $seq);
  my ($n, $slen, $comment, $qlen) = (0, 0, 0);
  while ( ($id, $seq, undef) = readfq($seqFh, \@aux)) {
 
    my $hash = &$hash_function($seq);
    # try for a right split so that only the last _ is used to split locus/allele
    #my ($locus, $allele) = split(/_/, $id);
    # https://stackoverflow.com/a/25173358
    my ($locus, $allele) = split(/_([^_]+)$/, $id);
 
    # Check if this is the first allele and if so, 
    # print out the sequence to reference alleles.
    # Increment the number of seqs
    if($numSeqs++ == 0){
      print $refFh ">$id\n$seq\n";
    }
    
    my $alleleLine = join("\t", $locus, $hash, $$settings{hash});

    print $allelesFh "$alleleLine\n";
  }

  return $numSeqs;
}

# Read fq subroutine from Andrea which was inspired by lh3
sub readfq {
    my ($fh, $aux) = @_;
    @$aux = [undef, 0] if (!(@$aux));	# remove deprecated 'defined'
    return if ($aux->[1]);
    if (!defined($aux->[0])) {
        while (<$fh>) {
            chomp;
            if (substr($_, 0, 1) eq '>' || substr($_, 0, 1) eq '@') {
                $aux->[0] = $_;
                last;
            }
        }
        if (!defined($aux->[0])) {
            $aux->[1] = 1;
            return;
        }
    }
    my $name = /^.(\S+)/? $1 : '';
    my $comm = /^.\S+\s+(.*)/? $1 : ''; # retain "comment"
    my $seq = '';
    my $c;
    $aux->[0] = undef;
    while (<$fh>) {
        chomp;
        $c = substr($_, 0, 1);
        last if ($c eq '>' || $c eq '@' || $c eq '+');
        $seq .= $_;
    }
    $aux->[0] = $_;
    $aux->[1] = 1 if (!defined($aux->[0]));
    return ($name, $seq) if ($c ne '+');
    my $qual = '';
    while (<$fh>) {
        chomp;
        $qual .= $_;
        if (length($qual) >= length($seq)) {
            $aux->[0] = undef;
            return ($name, $seq, $comm, $qual);
        }
    }
    $aux->[1] = 1;
    return ($name, $seq, $comm);
}

sub uc{
  return(uc($_[0]));
}

sub usage{
  print "$0: reads fasta file(s) and creates a database in the hashsum format
  The fasta file(s) must be one locus per line. The first allele is the assumed reference.
  Each sequence ID must have the format locus_allele.
  Usage: $0 [options] file1.fasta...
  --out    An output folder which will contain a reference fasta file and a TSV of alleles
  --hash   Which algorithm to use? md5 (default), sha256, sha1, plaintext
  --help   This useful help menu
  \n";
  exit 0;
}

File: scripts/digestFasta.py
#!/usr/bin/env python3
"""
reads fasta file(s) and creates a database in the hashsum format
The fasta file(s) must be one locus per line. The first allele is the assumed reference.
Each sequence ID must have the format locus_allele.

Input:
Fasta file

Output:
- A reference fasta file
- A TSV file with all the alleles
"""

import sys
import os
import argparse

import hashlib
import base64
def read_fasta(path):
        if path.endswith('.gz'):
            import gzip
            fasta = gzip.open(path, 'rt')
        else:
            fasta = open(path, 'rt')

        name = None
        
        for line in fasta:
            if line.startswith('>'):
                if name is not None:
                    yield name, seq
                name = line[1:].rstrip()
                seq = ''
            else:
                seq += line.rstrip()
        yield name, seq

def seq_digest(string, method="md5"):
    if method == "md5":
        return string.digest().encode('base64').strip()
    elif method == "md5_base64":
        return base64.b64encode(hashlib.md5(string.encode()).digest()).decode('utf-8').rstrip("=")
        
    else:
        raise ValueError("Unknown method %s" % method)

def main():
    args = argparse.ArgumentParser()
    args.add_argument("FASTA", help="Fasta file(s) with alleles having name as locus_id", nargs="+")
    args.add_argument("-o", "--out", "--outdir", help="Output directory, will be created")
    args.add_argument("--force", help="Proceed if output directory is found", action="store_true")
    args.add_argument("--verbose", help="Print verbose information", action="store_true")
    args = args.parse_args()

    locus_file = {}
    parsed_files = {}

    # TODO Check if output exists (force option)
    if os.path.isdir(args.out):
        if not args.force:
            print("Output directory exists, use --force to proceed", file=sys.stderr)
            return 1

    # Create output directory
    if not os.path.exists(args.out):
        os.makedirs(args.out)

    outputFastaRefs = open(os.path.join(args.out, "ref.fasta"), 'w')
    outputTsvAlleles = open(os.path.join(args.out, "alleles.tsv"), 'w')

    print("# locus", "allele", "hash-type", sep="\t", file=outputTsvAlleles)
    for fasta_path in args.FASTA:
        seq_num = 0
        if args.verbose:
            print("Processing %s" % fasta_path, file=sys.stderr)

        if not os.path.exists(fasta_path):
            print("File %s not found" % fasta_path, file=sys.stderr)
            return 1
        
        for name, seq in read_fasta(fasta_path):
            try:
                locus, allele = name.split("_")
            except ValueError:
                print("Error: %s is not a valid sequence name (should be locus_allele)" % name, file=sys.stderr)
                sys.exit(1)
            
            if locus not in locus_file:
                if args.verbose:
                    print(" + Locus %s from %s" % (locus, fasta_path), file=sys.stderr)
                locus_file[locus] = fasta_path
            elif locus_file[locus] != fasta_path:
                print("Locus %s is found in multiple files (%s and %s) as %s" % (locus, locus_file[locus], fasta_path, name), file=sys.stderr)
                return 1
            
            if seq_num == 0:
                print(">%s\n%s" % (name, seq), file=outputFastaRefs)
            
            # Encode MD5 base64 of seq
            base64_md5 = seq_digest(seq, method="md5_base64")
            print(locus, base64_md5, "md5",  sep="\t", file=outputTsvAlleles)
            seq_num += 1

if __name__ == "__main__":
    quit(main())
